### ISSUE TITLE
Off-Topic, but I am desperate: PE-Code Manager Web Hooks do not work

### ISSUE THREAD

**Issue by linuxdan** [voxpupuli/puppet-gitlab#327]
PE 2019.1 running on CentOS 7
GitLab 12.4.0-ee running on CentOS 7

I followed https://puppet.com/docs/pe/2019.1/code_mgr_webhook.html#triggering-code-manager-with-a-webhook to set it up.  I disabled SSL verification and enabled the "Allow outbound requests to the local network" settings
 
I created an authentication token and set up the URL as directed:

`https://code-manager.example.com:8170/code-manager/v1/webhook?type=gitlab&token=<TOKEN>`

HOWEVER, when I test it from GitLab, I get an OOPS 500 error on the GUI and an 
error in the gitlab-rails/production.log saying 

`Net::HTTPFatalError (503 "Service Unavailable")`

 If I change the "https" to "http", the test completes, but I get a 404 
Network Error in the response from Puppet to GitLab, saying that the 
hostname of the Puppet Server could not be resolved by DNS.  The full FQDN 
is in both DNS and the hosts files of both servers.

If I actually push code into the git project, I get the same results on the GitLab end, and absolutely nothing on the Puppet end.

However, `puppet-code deploy ...` works perfectly.

Any clues ?

**Comment by LongLiveCHIEF** [comment]
Make sure your DNS host can resolve both FQDN's from each server would be my initial instinct. Also make sure each host (gitlab and puppet code manager) both use the DNS host you expect.

Are these services on a public network? Internet network? same subnet? Just because the full FQDN is in the hosts files, doesn't mean that the actual IP address can be resolved if they are on different broadcast networks.

I'm closing because this isn't related to this module and I don't want it to confuse anyone, but I'll continue to toss around ideas here in the thread.

**Comment by linuxdan** [comment]
Thanks for your indulgence.  I am getting no traction on the Puppet User's mailing list or on the Slack channels.
This is a semi-private test-lab network where we run our own little DNS.
I tested lookups with `dig` and `getent hosts` and it all works on both servers for both servers.

**Comment by LongLiveCHIEF** [comment]
Any firewall rules in place?

**Comment by LongLiveCHIEF** [comment]
I'm wondering if Gitlab is trying to use public DNS, instead of whatever you have configured in `/etc/resolv.conf`. I seem to remember an issue on this a few years ago, and I believe they added support for a configure to allow you to specify IP address of DNS host.

I'll do some digging, but that's my current hunch and just wanted to let you know where I'm at on this.

**Comment by LongLiveCHIEF** [comment]
Also, just as a double-check, make sure if you do have firewall, that you're allowing outgoing tcp/53.

**Comment by LongLiveCHIEF** [comment]
And another thought... what happens if you use IP addresses in stead of hostnames just as a check?  Let's see if we can make sure the problem is application using it's own DNS configs.

(sorry for the spam! lots of interest in this problem, even though this is first I've heard of code manager)

**Comment by linuxdan** [comment]
I think I disabled all proxy settings and firewalls.  I can check.
I will also try using the IP address in the hook.

I am taking advantage of the joys of virtual machines to stand up a new pair from scratch.
If GitLab does lookups that bypass nsswitch.conf, I may be skroood.  This is targeted for a production environment where DNS is used sparingly for security reasons.

**Comment by LongLiveCHIEF** [comment]
if they do lookups that don't pull in resolv.conf, then that means they are
configurable via the gitlab.rb. It just might not be one of the default
templated keys you see and we'd have to dig to the underlaying configs that
are generated by that file to find they key setting for this.

**Comment by linuxdan** [comment]
That 404/DNS failure could also be coming from Code Manager’s listener on port 8170
I need to see if that can be traced / debugged

**Comment by linuxdan** [comment]
> (sorry for the spam! lots of interest in this problem, even though this is first I've heard of code manager)

Dude, seriously ? You have never tinkered with the Enterprise unique features ?
It is never too late to start   😁

**Comment by linuxdan** [comment]
New data point !
All proxy settings are unset.
Firewalls at both ends disabled.
Test returns the same failure/errors.

Now I replace the host name in the hook URL with the IP address.
https still gives the 500/503, but http retutns a 403 - Access Denied !
The response body contains:
Access Denied (policy_denied)
Your system policy has denied access to the requested URL

Searching Puppet logs, I find no instance of "denied", but policy turns up in console-services, orchestration-services, puppetdb, and puppetserver logs 
The log lines are INFO and they all say 

`[p.t.s.m.jolokia] Using policy access restrictor classpath:/jolokia-access.xml`

**Comment by LongLiveCHIEF** [comment]
what is the visibility of the repo with the hook?

**Comment by linuxdan** [comment]
Internal.
HOWEVER !!  I just finished standing up a new GitLab (CE-12.4.2) along with a new PE and they work !!
First smoking gun test: I tried using the new PE and the old GitLab - and GitLab barfed !

By the way, I had issues trying to install PE-2019.2 and fell back to PE-2019.1.1
How do I report it when I am still waiting for a proper PE license ?

**Comment by LongLiveCHIEF** [comment]
I don't know anything about PE or the admin side of puppet really. I'm just
a lowly module developer. I publish to forge and that's where it ends for
me.

**Comment by linuxdan** [comment]
I never heard a peep from the GitLab community other than, "Did you try all the obvious stuff in the documentation ?"

The PE install problem was that IPv6 was disabled and I did not know I had to tell the installer about it :grin: My thanks to the Puppet Slack channel for that.

I have no clue what I did differently on the new GitLab, but I was able to bump it up to the latest-greatest EE version.

Thanks for your feedback and willingness to help.
